File: competitive-programming/leet-code/210/findOrder.js
var findOrder = function (numCourses, prerequisites) {

    const adj = Array.from(
        { length: numCourses },
        () => []
    );

    const indegree = new Array(numCourses).fill(0);

    for (const [course, prereq] of prerequisites) {
        adj[prereq].push(course);
        indegree[course]++;
    }

    const q = [];

    for (let i = 0; i < numCourses; i++) {
        if (indegree[i] === 0) {
            q.push(i);
        }
    }

    const ans = [];

    while (q.length) {

        const curr = q.shift();
        ans.push(curr);

        for (const neig of adj[curr]) {

            indegree[neig]--;

            if (indegree[neig] === 0) {
                q.push(neig);
            }
        }
    }

    return ans.length === numCourses ? ans : [];
};

File: competitive-programming/leet-code/210/find_order.py
from typing import List
from collections import deque

def find_order(num_courses: int, prerequisites: List[List[int]]) -> List[int]:
    adj = [[] for _ in range(num_courses)]
    indegree = [0] * num_courses

    for course, prereq in prerequisites:
        adj[prereq].append(course)
        indegree[course] += 1

    q = deque([i for i in range(num_courses) if indegree[i] == 0])

    ans = []

    while q:
        curr = q.popleft()
        ans.append(curr)

        for neighbor in adj[curr]:
            indegree[neighbor] -= 1
            if indegree[neighbor] == 0:
                q.append(neighbor)

    return ans if len(ans) == num_courses else []